daemon/src/common/Assert.h

```diff
@@ -0,0 +1,159 @@
+        #define DAEMON_ASSERT_CALLSITE_HELPER(file, func, line, directExpression, _1, _2, _3)
+    #endif
+#endif
+
+/*
+ * You can define new ASSERT variants by using DAEMON_ASSERT_CALLSITE:
+ *     - directExpression should be the condition expression, evaluating to a bool. It is used
+ *       when hinting the compiler optimizer
+ *     - code should be used to defined variables that will appear in condition and message. Operands
+ *       should be evaluated once here using "auto&& var = (op)" so that functions are not called
+ *       multiple times
+ *     - condition is the same expression as directExpression but using variables defined in code instead
+ *     - message should evaluate to a string explaining the error and dumping the variables defined in code.
+ */
+#define DAEMON_ASSERT_CALLSITE(directExpression, code, condition, message) \
+    DAEMON_ASSERT_CALLSITE_HELPER(__FILE__, __func__, __LINE__, directExpression, code, condition, message)
+
+#define DAEMON_ASSERT(condition) DAEMON_ASSERT_CALLSITE( \
+        (condition), \
+        , \
+        (condition), \
+        Str::Format("\"%s\" is false", #condition) \
+    )
+
+#define DAEMON_ASSERT_EQ(a, b) DAEMON_ASSERT_CALLSITE( \
+        (a) == (b), \
+        auto&& expected = (a); auto&& actual = (b), \
+        expected == actual, \
+        Str::Format("\"%s == %s\" expected: %s, actual: %s", #a, #b, expected, actual) \
+    );
+#define DAEMON_ASSERT_NQ(a, b) DAEMON_ASSERT_CALLSITE( \
+        (a) != (b), \
+        auto&& notExpected = (a); auto&& actual = (b), \
+        notExpected != actual, \
+        Str::Format("\"%s != %s\" not expected: %s, actual: %s", #a, #b, notExpected, actual) \
+    );
+
+#define DAEMON_ASSERT_LT(a, b) DAEMON_ASSERT_CALLSITE( \
+        (a) < (b), \
+        auto&& bound = (a); auto&& actual = (b), \
+        bound < actual, \
+        Str::Format("\"%s < %s\" bound: %s, actual: %s", #a, #b, a, b) \
+    );
+#define DAEMON_ASSERT_LE(a, b) DAEMON_ASSERT_CALLSITE( \
+        (a) <= (b), \
+        auto&& bound = (a); auto&& actual = (b), \
+        bound <= actual, \
+        Str::Format("\"%s <= %s\" bound: %s, actual: %s", #a, #b, a, b) \
+    );
+#define DAEMON_ASSERT_GT(a, b) DAEMON_ASSERT_CALLSITE( \
+        (a) > (b), \
+        auto&& bound = (a); auto&& actual = (b), \
+        bound > actual, \
+        Str::Format("\"%s > %s\" bound: %s, actual: %s", #a, #b, a, b) \
+    );
+#define DAEMON_ASSERT_GE(a, b) DAEMON_ASSERT_CALLSITE( \
+        (a) >= (b), \
+        auto&& bound = (a); auto&& actual = (b), \
+        bound >= actual, \
+        Str::Format("\"%s >= %s\" bound: %s, actual: %s", #a, #b, a, b) \
+    );
+
+#if !defined(DAEMON_SKIP_ASSERT_SHORTHANDS)
+    #define ASSERT DAEMON_ASSERT
+    #define ASSERT_EQ DAEMON_ASSERT_EQ
+    #define ASSERT_NQ DAEMON_ASSERT_NQ
+    #define ASSERT_LT DAEMON_ASSERT_LT
+    #define ASSERT_LE DAEMON_ASSERT_LE
+    #define ASSERT_GT DAEMON_ASSERT_GT
+    #define ASSERT_GE DAEMON_ASSERT_GE
+#endif
```

Comment:
ASSERT only tells you that the values are different, ASSERT_EQ also tells you what the values are.